File: LeetCode/p46.cpp
class Solution {
public:
    vector<vector<int>> permute(const vector<int> &nums) {
        raw_data = nums;
        len = nums.size();
        used = vector<bool>(len);
        backtracking();
        return result;
    }

private:
    int len;
    vector<int> raw_data;
    vector<bool> used;
    vector<int> tmp_vec;
    vector<vector<int>> result;

    void backtracking() {
        if (tmp_vec.size() == len) {
            result.emplace_back(tmp_vec);
            return;
        }
        for (int i = 0; i < len; ++i) {
            if (used[i])
                continue;
            tmp_vec.push_back(raw_data[i]);
            used[i] = true;
            backtracking();
            used[i] = false;
            tmp_vec.pop_back();
        }
    }
};

File: LeetCode/p46.rs
struct Combine {
  nums: Vec<i32>,
  tmp: Vec<i32>,
  used: Vec<bool>,
  ret: Vec<Vec<i32>>,
}

impl Combine {
  pub fn combine(mut self) -> Vec<Vec<i32>> {
    self.backtrack();
    self.ret
  }

  fn backtrack(&mut self) {
    if self.tmp.len() == self.nums.len() {
      self.ret.push(self.tmp.clone());
      return;
    }
    for i in 0..self.nums.len() {
      if self.used[i] {
        continue;
      }
      self.tmp.push(self.nums[i]);
      self.used[i] = true;
      self.backtrack();
      self.used[i] = false;
      self.tmp.pop();
    }
  }
}

impl Solution {
  pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
    Combine{used: vec![false; nums.len()], nums, tmp: vec![], ret: vec![]}.combine()
  }
}

